### PR TITLE
Summer Thaw: The Unused Ideas 1: Progress Bar for interaction and do_after()

### PR DESCRIPTION
**Pull request by Iamgoofball**
# WITH THE NEW UPDATE, IT IS NOW CLIENTSIDE. OTHER PEOPLE CANNOT SEE YOUR PROGRESS BAR.

animationfinal

a variant upon https://tgstation13.org/phpBB/viewtopic.php?f=5&t=3771#p93276
basically if you do any action where you gotta stand still for a bit, this makes a progress bar show up

every other multiplayer game has something like this if it has interaction of this kind

since animations of the player are out of the question, this works best for the situation

this way if you're welding off a wall, people know not to bump into you.

### CODE REVIEW

**Comment by Dorsisdwarf** [comment]
Sprite could be reworked to be in HUD colors, perhaps?

**Comment by Fox-McCloud** [comment]
Goon has this, as well, but it ONLY appears for the plaeyr doing the action---in part to allow for a bit more stealthy approaches to thing.

Either case It's...an interesting feature, so I have mixed feelings about it.

**Comment by Jalleo** [comment]
could do it for actions when in non harm modes only

**Comment by MrPerson** [comment]
Down the line it would be nice to have the image set per-action or at least per-object, but for now this is an ok framework to expand on.

Should prob be changelog'd.

**Comment by Iamgoofball** [comment]
@fox-mccloud odd, usually I can see it when others do it

kinda kills the point of it if only you can see it

**Comment by Fox-McCloud** [comment]
@Iamgoofball 

I think there's a lot of inconsistency---intentionally, in a lot of cases.

**Comment by Iamgoofball** [comment]
@Fox-McCloud 
APCs taking time to emag is rather dumb

removing cuffs in front of anyone is a bad idea, find somewhere to hide to do that

**Comment by xxalpha** [comment]
God no. :-1:

**Comment by Iamgoofball** [comment]
@xxalpha why

provide reasoning

**Comment by Iamgoofball** [comment]
new sprites in OP btw

**Comment by Fox-McCloud** [comment]
Now that is MUCH better.

**Comment by xxalpha** [comment]
Looks awful and out of place. Not necessary. Takes up a lot of space. Unnecessary display of information.

**Comment by samis** [comment]
This looks nice, but wouldn't it work best if it was only visible to the player doing the action?

**Comment by optimumtact** [comment]
This should be visible to the player only, other players do not need to see when you are doing actions

**Comment by xxalpha** [comment]
>This should be visible to the player only, other players do not need to see when you are doing actions

This is critical, but I still think it's bad.

**Comment by xxalpha** [comment]
Make it visible to the player only and make it a preference so I can disable it.

**Comment by Anonus** [comment]
Definitely make it visible to the player only, or it'll get annoying real fast.
Also, it should probably appear on the object that you're doing the work on instead of just near it.

**Comment by Iamgoofball** [comment]
I've got a thing I'm working on to make it much much much less intrusive and obnoxious.

**Comment by Iamgoofball** [comment]
Also, the thing is, one of the main reasons I'm adding this is so that you can tell if another player is sitting there idling or if he's unwrenching something/welding down a wall/screwing in a dong and how long it's gonna take them to finish. I want to keep it visible for everyone around.

The new version is much less intrusive, I'll put up a gif soon.

**Comment by xxalpha** [comment]
>I want to keep it visible for everyone around.

No, that's awful.

**Comment by Iamgoofball** [comment]
Wait till you see the new version.

**Comment by xxalpha** [comment]
You don't understand. This should not be visible to everyone.

**Comment by Iamgoofball** [comment]
Explain why it shouldn't be visible to everyone.

**Comment by xxalpha** [comment]
Visual clutter, stealth, looks out of place.

**Comment by Iamgoofball** [comment]
>visual clutter
Wait for the next gif to go up.
>stealth
If we want certain interactions to be stealthy, I can just add a var to do_after() that if you enable it, it becomes stealthy.
>looks out of place
Wait for the next gif to go up.

**Comment by Iamgoofball** [comment]
New GIFs uploaded, check the OP

**Comment by ike709** [comment]
I do like this alot, but I also think preferences for this should exist. Show Everyone's, Show Yours, or Hide All. Not disabling any to other people, just keeping certain ones invisible to you specifically based on the selected preference, if that's possible without speshul snowflakes and whatnot.

**Comment by Fox-McCloud** [comment]
This is pretty much like Goon's, at this stage, minus the icons to indicate what's going on---it's also  a lot nicer looking progress bar and far far less intrusive.

Do agree there should be a stealth var for "just in case" (can't think of anything that would need it at the moment though).

**Comment by Iamgoofball** [comment]
I've done all I can to it at this point.

Someone else will need to make it visible only to the owner, I guess.

Or make a preference or something

**Comment by PKPenguin321** [comment]
>Goon has this, as well, but it ONLY appears for the plaeyr doing the action---in part to allow for a bit more stealthy approaches to thing.

This
I've had moments as antag when I was locked to a chair in the brig and was barely able to slip out of my cuffs and make a mad dash out without the guards noticing. One moment in particular was highly memorable when I not only did this, but hid under another person on a pod to hide from security afterwords. This would not have happened if the guards had a **BIG FLASHY PROGRESS BAR** to check if I was breaking out or not. Have it just appear to you instead of to everyone (or make it only appear on help intent) so you can still do stealthy things.

**Comment by HoaxIT** [comment]
:-1: not using smooth animations

**Comment by SconesC** [comment]
Should be visible to the player only

**Comment by Iamgoofball** [comment]
@HoaxIT there is literally no way to do a smooth bar animation here. I can only update the bar every tick.

**Comment by Jordie0608** [comment]
An alternative display could be to change mouse icon into a spinning timer or whatever by fiddling with the pointer icon; however it'd be only visible to the user of course.

**Comment by Iamgoofball** [comment]
>fucking with the cursor

only if we add cursors for shift-click examining, when you have throw mode on, etc.

**Comment by Jordie0608** [comment]
I'm going to be fiddling with the cursor once 508 leaves beta and while I probably won't do that yet, it's pretty easy to do.

**Comment by Fox-McCloud** [comment]
I'll admit, that as much as this may make certain actions obvious...I dunno if a lot of those actions should be non-obvious because...well...someone breaking down a wall or breaking out of cuffs is generally going to be...decently obvious if you were to walk into the same room as them.

**Comment by Iamgoofball** [comment]
@HoaxIT 
animation7
this is as smooth as we're getting dude, sorry

**Comment by Iamgoofball** [comment]
# WITH THE NEW UPDATE, IT IS NOW CLIENTSIDE. OTHER PEOPLE CANNOT SEE YOUR PROGRESS BAR.

**Comment by Miauw62** [comment]
add an argument to do_after() so coders can decide whether they want other people to be able to see the bar or not.

**Comment by Iamgoofball** [comment]
@Miauw62 it's now 100% client.images though

**Comment by xxalpha** [comment]
Great. Now all that is left to do is make it a preference.

**Comment by Iamgoofball** [comment]
@xxalpha preference for what?

**Comment by xxalpha** [comment]
Turn these on and off.

**Comment by Iamgoofball** [comment]
@xxalpha no

**Comment by HoaxIT** [comment]
@Iamgoofball sexy.

**Comment by RemieRichards** [comment]
@Iamgoofball yes, but you can still add the argument dumb dumb! just have a check for the argument before it puts it in ```client.images``` or ```user.overlays``` silly.

File: code/__HELPERS/unsorted.dm [hunk 1]
```diff
@@ -683,13 +683,47 @@ Turf and target are seperate in case you want to teleport some distance from a t
...
+					progbar.icon_state = "prog_bar_70"
+				if(81 to 90)
+					progbar.icon_state = "prog_bar_80"
+				if(91 to INFINITY)
+					progbar.icon_state = "prog_bar_90"
+			progbar.pixel_y = 32
+			user.client.images |= progbar
 		sleep(timefraction)
 		if(!user || !target)
+			if(user && user.client)
```
Comment (RemieRichards) [inline code/__HELPERS/unsorted.dm]: this can just be ```if(user.client)``` the above check covers it.

File: code/__HELPERS/unsorted.dm [hunk 2]
```diff
@@ -683,13 +683,47 @@ Turf and target are seperate in case you want to teleport some distance from a t
...
+				user.client.images -= progbar
 			return 0
 		if ( user.loc != user_loc || target.loc != target_loc || user.get_active_hand() != holding || user.incapacitated() || user.lying )
+			if(user && user.client)
```
Comment (RemieRichards) [inline code/__HELPERS/unsorted.dm]: this can just be ```if(user.client)``` the above check covers it

**Comment by d3athrow** [comment]
Benis

**Comment by Anonus** [comment]
Put the image on the thing you're interacting with, so you can see how long multiple objects have until completion.

**Comment by duncathan** [comment]
How does this work if you're doing multiple actions at once? Like welding three walls.

**Comment by Iamgoofball** [comment]
animationfinal

BAR NOW SHOWS UP ON THE OBJECT YOU'RE INTERACTING WITH 

@duncathan

**Comment by nukeop** [comment]
Merge ASAP.

Is this also enabled for longer actions such as unbuckling/removing handcuffs?

**Comment by Anonus** [comment]
Now it's :+1:

**Comment by Iamgoofball** [comment]
@nukeop yes

**Comment by optimumtact** [comment]
Looks excellent now :+1:

**Comment by freerealestate** [comment]
:+1: Impressive

**Comment by Fox-McCloud** [comment]
should still probably be a preference.

**Comment by Fox-McCloud** [comment]
Goooofff.

Preferenceeee optionnnnn.

File: code/__HELPERS/unsorted.dm [hunk 3]
```diff
@@ -683,15 +683,34 @@ Turf and target are seperate in case you want to teleport some distance from a t
 	var/target_loc = target.loc
 	var/holding = user.get_active_hand()
 	var/timefraction = round(time/numticks)
-	for(var/i = 0, i<numticks, i++)
+	var/image/progbar
+	for(var/i = 1 to numticks)
```
Comment (optimumtact) [inline code/__HELPERS/unsorted.dm]: Put this part inside another internal helper proc, then you can do the progbar bits around the outside of the method

**Comment by Razharas** [comment]
Whats the difference between do_mob() and do_after()?

**Comment by Iamgoofball** [comment]
One tracks the locations of 2 mobs, 1 tracks the location of 1 mob

**Comment by RemieRichards** [comment]
@Iamgoofball Except ```do_after()``` has a /atom/target argument now, and what are mobs? /atom/movable, One of the two is clearly obsolete now (hint, it's ```do_mob```!)

**Comment by optimumtact** [comment]
If I recall I did intend to merge the two when I added the target argument, but there was something that stopped me, although of course I cannot remember what it is now!

Anyway yeah, do_mob could just call and return do_after with the right parameters.

**Comment by RemieRichards** [comment]
@optimumtact or we could just not be lazy, update all calls of ```do_mob``` to ```do_after``` and avoid having ```do_mob``` give us double overhead, k?

**Comment by optimumtact** [comment]
I don't think a double function call should be very much overhead

If you want to go through and update every single call to do_mob be my guest, it's unfair to expect goof to do that however.

**Comment by RemieRichards** [comment]
@optimumtact I'm not expecting goof too at all, I never stated that anywhere and if it came across as such it was not my intention.

Yet the topic of "why do we have two procs that used to be fairly similair which one of the two now completely encompasses the other's functionality" was brought up, and I offered what I know on said topic.

**Comment by optimumtact** [comment]
Fair enough.

**Comment by Fox-McCloud** [comment]
..Still think this needs to be a user preference option.

**Comment by TheVekter** [comment]
I don't think anyone is gonna say "I really don't want to know exactly when
I'm going to finish wrenching this wall".

**Comment by Niknakflak** [comment]
I really don't want to know exactly when I'm going to finish wrenching this wall.

**Comment by TheVekter** [comment]
May I ask why? Seems like a pretty universally desired addition.

**Comment by Dorsisdwarf** [comment]
Why does this need a preference option to fill up the already pointlessly bloated preferences screen, filled by people bleating that they don't want any QOL features at all?

**Comment by Iamgoofball** [comment]
I'm not gonna do a preference option then I guess

**Comment by orelbon** [comment]
:-1:  they look pretty bad and ruin the atmosphere of the game for me, unless you made them optional then i say no.

**Comment by Iamgoofball** [comment]
@orelbon 
>ss13
>atmosphere

**Comment by orelbon** [comment]
lol i actualy like ze atmoshpere

**Comment by Iamgoofball** [comment]
regardless this PR has been merged and live on the server for 8 days

**Comment by orelbon** [comment]
honestly i dont hate it i just dont like the idea of having progress bars on an RP server. yes i know paradise! RP?

**Comment by Fox-McCloud** [comment]
@orelbon 

This is TG Station, son.

**Comment by Iamgoofball** [comment]
@orelbon 

gameplay > RP

always

**Comment by optimumtact** [comment]
tgstation

**Comment by orelbon** [comment]
actually i just say the client side part...so now since only we can see our bars i'm good with it.

**Pull request merged**